src/java/org/apache/cassandra/auth/CassandraAuthorizer.java

```diff
@@ -92,18 +96,41 @@ public CassandraAuthorizer()
         }
     }
 
-    public void grant(AuthenticatedUser performer, Set<Permission> permissions, IResource resource, RoleResource grantee)
+    public Set<Permission> grant(AuthenticatedUser performer, Set<Permission> permissions, IResource resource, RoleResource grantee)
     throws RequestValidationException, RequestExecutionException
     {
-        modifyRolePermissions(permissions, resource, grantee, "+");
-        addLookupEntry(resource, grantee);
+        String roleName = escape(grantee.getRoleName());
+        String resourceName = escape(resource.getName());
+        Set<Permission> grantedPermissions = Sets.newHashSetWithExpectedSize(permissions.size());
+        Set<Permission> nonExistingPermissions = new HashSet<>(permissions);
+
+        nonExistingPermissions.removeAll(getExistingPermissions(roleName, resourceName, permissions));
+        if (!nonExistingPermissions.isEmpty())
+        {
+            modifyRolePermissions(nonExistingPermissions, resource, grantee, "+");
+            addLookupEntry(resource, grantee);
+            grantedPermissions.addAll(nonExistingPermissions);
+        }
+
+        return grantedPermissions;
```

Comment:
I think `grantedPermissions` is redundant, since it's an exact copy of `nonExistingPermissions`. We can probably simplify to something like:
```suggestion
        Set<Permission> existingPermissions = getExistingPermissions(roleName, resourceName, permissions);
        Set<Permission> nonExistingPermissions = Sets.difference(permissions, existingPermissions);

        if (!nonExistingPermissions.isEmpty())
        {
            modifyRolePermissions(nonExistingPermissions, resource, grantee, "+");
            addLookupEntry(resource, grantee);
        }

        return nonExistingPermissions;
```